apps/livreGUI/transferFunctionEditor/RangeWidget.cpp

```diff
@@ -0,0 +1,182 @@
+             });
+    connect( this, &RangeWidget::rangeXChanged, [&](qreal value)
+            {
+                if( _histogram.isEmpty( ))
+                    return;
+                value = _convertFromNormalizedRange( value );
+                _ui->minDataRange->setMinimum( std::min( _ui->minDataRange->minimum(), double( value )));
+                _ui->minDataRange->setValue( value );
+            });
+
+    auto leftRange = _ui->background->rootObject()->findChild<QObject*>("leftrange");
+    connect( leftRange, SIGNAL(viewposChanged(qreal)), this, SLOT(setRangeX(qreal)));
+
+
+    connect( _ui->maxDataRange,
+             static_cast<void (QDoubleSpinBox::*)(double)>(&QDoubleSpinBox::valueChanged),
+             [&] ( const double value )
+             {
+                _ui->minDataRange->setMaximum( value );
+                if( !_histogram.isEmpty( ))
+                    setRangeY( _convertToNormalizedRange( value ));
+             });
+    connect( this, &RangeWidget::rangeYChanged, [&](qreal value)
+            {
+                if( _histogram.isEmpty( ))
+                    return;
+                value = _convertFromNormalizedRange( value );
+                _ui->maxDataRange->setMaximum( std::max( _ui->maxDataRange->maximum(), double( value )));
+                _ui->maxDataRange->setValue( value );
+            });
+
+    auto rightRange = _ui->background->rootObject()->findChild<QObject*>("rightrange");
+    connect( rightRange, SIGNAL(viewposChanged(qreal)), this, SLOT(setRangeY(qreal)));
+}
+
+RangeWidget::~RangeWidget()
+{
+    delete _ui;
+}
+
+void RangeWidget::setHistogram( const lexis::render::Histogram& histogram,
+                                const bool isLogScale )
+{
+    if( histogram == _histogram && isLogScale == _isLogScale )
+        return;
+
+    const bool haveHistogram = !histogram.isEmpty();
+    _ui->minDataRange->setEnabled( haveHistogram );
+    _ui->maxDataRange->setEnabled( haveHistogram );
+    _ui->fitToDataButton->setEnabled( haveHistogram );
+
+    _histogram = histogram;
+    _isLogScale = isLogScale;
+
+    emit rangeXChanged( _rangeX );
+    emit rangeYChanged( _rangeY );
+
+    _ui->background->rootObject()->setProperty( "histogram",
+                                                sampleHistogram( _histogram, _isLogScale ));
+}
+
+bool RangeWidget::eventFilter( QObject* watched, QEvent* event )
+{
+    if( watched == _ui->background && event->type() == QEvent::MouseMove )
+    {
+        if(  _histogram.isEmpty( ))
+            return true;
+
+        QMouseEvent* mouseEvent = static_cast< QMouseEvent* >( event );
+
+        const size_t index = (float)_histogram.getBins().size() *
+                             (float)mouseEvent->pos().x() / (float)width();
+        emit histIndexChanged( index, _histogram.getRatio( index ));
+    }
+    return false;
+}
+
+void RangeWidget::setRangeX(qreal rangeX)
+{
+    if( std::abs( _rangeX - rangeX ) <= 0.0001 )
```

Comment:
Yes, it's also good to know that it's used internally by the QRectF / QsizeF == operators